File: gateway/ui/src/ui/gateway-client.ts
/**
 * Gateway WebSocket Client
 * 
 * Handles connection to GSV Gateway via WebSocket.
 * Matches the protocol used by the Rust CLI.
 */

import type { Frame, RequestFrame, ResponseFrame, EventFrame, ToolDefinition } from "./types";

export type ConnectionState = "disconnected" | "connecting" | "connected";

export type GatewayClientOptions = {
  url: string;
  token?: string;
  onStateChange?: (state: ConnectionState) => void;
  onEvent?: (event: EventFrame) => void;
  onError?: (error: string) => void;
};

type PendingRequest = {
  resolve: (response: ResponseFrame) => void;
  reject: (error: Error) => void;
  timeout: ReturnType<typeof setTimeout>;
};

export class GatewayClient {
  private ws: WebSocket | null = null;
  private pending = new Map<string, PendingRequest>();
  private reconnectAttempts = 0;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  
  private _state: ConnectionState = "disconnected";
  private options: GatewayClientOptions;

  constructor(options: GatewayClientOptions) {
    this.options = options;
  }

  get state(): ConnectionState {
    return this._state;
  }

  private setState(state: ConnectionState) {
    this._state = state;
    this.options.onStateChange?.(state);
  }

  /**
   * Start connection to Gateway
   */
  start(): void {
    if (this.ws) return;
    this.connect();
  }

  /**
   * Stop connection
   */
  stop(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.setState("disconnected");
  }

  private connect(): void {
    this.setState("connecting");
    
    try {
      this.ws = new WebSocket(this.options.url);
      
      this.ws.onopen = () => {
        console.log("[GatewayClient] Connected");
        this.reconnectAttempts = 0;
        this.handshake();
      };
      
      this.ws.onmessage = (event) => {
        this.handleMessage(event.data);
      };
      
      this.ws.onclose = () => {
        console.log("[GatewayClient] Disconnected");
        this.ws = null;
        this.setState("disconnected");
        this.scheduleReconnect();
      };
      
      this.ws.onerror = (error) => {
        console.error("[GatewayClient] Error:", error);
        this.options.onError?.("Connection failed");
      };
    } catch (e) {
      console.error("[GatewayClient] Failed to connect:", e);
      this.options.onError?.(`Failed to connect: ${e}`);
      this.scheduleReconnect();
    }
  }

  private scheduleReconnect(): void {
    if (this.reconnectTimer) return;
    
    const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts), 30000);
    this.reconnectAttempts++;
    
    console.log(`[GatewayClient] Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts})`);
    
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this.connect();
    }, delay);
  }

  private async handshake(): Promise<void> {
    try {
      const response = await this.request("connect", {
        minProtocol: 1,
        maxProtocol: 1,
        client: {
          id: `web-${crypto.randomUUID().slice(0, 8)}`,
          version: "0.1.0",
          platform: "web",
          mode: "client",
        },
        auth: this.options.token ? { token: this.options.token } : undefined,
      });
      
      if (!response.ok) {
        const errorMsg = response.error?.message || "Handshake failed";
        this.options.onError?.(errorMsg);
        throw new Error(errorMsg);
      }
      
      console.log("[GatewayClient] Handshake complete");
      this.setState("connected");
    } catch (e) {
      console.error("[GatewayClient] Handshake failed:", e);
      const errorMsg = e instanceof Error ? e.message : String(e);
      this.options.onError?.(errorMsg);
      this.ws?.close();
    }
  }

  private handleMessage(data: string): void {
    let frame: Frame;
    try {
      frame = JSON.parse(data);
    } catch {
      console.error("[GatewayClient] Invalid JSON:", data);
      return;
    }

    if (frame.type === "res") {
      const pending = this.pending.get(frame.id);
      if (pending) {
        clearTimeout(pending.timeout);
        this.pending.delete(frame.id);
        pending.resolve(frame);
      }
    } else if (frame.type === "evt") {
      this.options.onEvent?.(frame);
    }
  }

  /**
   * Send a request and wait for response
   */
  async request<T = unknown>(method: string, params?: unknown, timeoutMs = 30000): Promise<ResponseFrame & { payload?: T }> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error("Not connected");
    }

    const id = crypto.randomUUID();
    const frame: RequestFrame = {
      type: "req",
      id,
      method,
      params,
    };

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error(`Request timeout: ${method}`));
      }, timeoutMs);

      this.pending.set(id, { resolve: resolve as (r: ResponseFrame) => void, reject, timeout });
      this.ws!.send(JSON.stringify(frame));
    });
  }

  // ---- Convenience methods ----

  async chatSend(sessionKey: string, message: string, runId?: string): Promise<ResponseFrame> {
    return this.request("chat.send", {
      sessionKey,
      message,
      runId: runId || crypto.randomUUID(),
    });
  }

  async sessionsList(limit = 100, offset = 0): Promise<ResponseFrame> {
    return this.request("sessions.list", { limit, offset });
  }

  async sessionGet(sessionKey: string): Promise<ResponseFrame> {
    return this.request("session.get", { sessionKey });
  }

  async sessionStats(sessionKey: string): Promise<ResponseFrame> {
    return this.request("session.stats", { sessionKey });
  }

  async sessionReset(sessionKey: string): Promise<ResponseFrame> {
    return this.request("session.reset", { sessionKey });
  }

  async sessionCompact(sessionKey: string, keepMessages = 20): Promise<ResponseFrame> {
    return this.request("session.compact", { sessionKey, keepMessages });
  }

  async sessionPreview(sessionKey: string, limit = 50): Promise<ResponseFrame> {
    return this.request("session.preview", { sessionKey, limit });
  }

  async toolsList(): Promise<ResponseFrame> {
    return this.request("tools.list");
  }

  async toolInvoke(tool: string, args: Record<string, unknown>): Promise<ResponseFrame> {
    return this.request("tool.invoke", { tool, args }, 60000);
  }

  async configGet(path?: string): Promise<ResponseFrame> {
    return this.request("config.get", path ? { path } : undefined);
  }

  async configSet(path: string, value: unknown): Promise<ResponseFrame> {
    return this.request("config.set", { path, value });
  }

  async channelsList(): Promise<ResponseFrame> {
    return this.request("channels.list");
  }

  async channelStart(channel: string, accountId = "default", config?: Record<string, unknown>): Promise<ResponseFrame> {
    return this.request("channel.start", { channel, accountId, config });
  }

  async channelStop(channel: string, accountId = "default"): Promise<ResponseFrame> {
    return this.request("channel.stop", { channel, accountId });
  }

  async channelStatus(channel: string, accountId = "default"): Promise<ResponseFrame> {
    return this.request("channel.status", { channel, accountId });
  }
}

File: cli/src/gateway_client.rs
use crate::connection::Connection;
use serde::Serialize;
use serde_json::{json, Map, Value};

pub type GatewayResult<T> = Result<T, Box<dyn std::error::Error>>;

pub struct GatewayClient {
    conn: Connection,
}

impl GatewayClient {
    pub fn new(conn: Connection) -> Self {
        Self { conn }
    }

    pub async fn connect(url: &str, token: Option<String>) -> GatewayResult<Self> {
        let conn = Connection::connect_with_options(url, "client", None, None, |_| {}, None, token)
            .await?;

        Ok(Self::new(conn))
    }

    async fn request<TParams: Serialize>(
        &self,
        method: &'static str,
        params: Option<TParams>,
    ) -> GatewayResult<Value> {
        let params = params.map(serde_json::to_value).transpose()?;
        let response = self.conn.request(method, params).await?;

        if !response.ok {
            let message = response
                .error
                .as_ref()
                .map(|error| format!("{} (code {}): {}", method, error.code, error.message))
                .unwrap_or_else(|| format!("{} failed", method));

            return Err(message.into());
        }

        Ok(response.payload.unwrap_or_else(|| json!({})))
    }

    pub fn connection(&self) -> &Connection {
        &self.conn
    }

    pub async fn heartbeat_status(&self) -> GatewayResult<Value> {
        self.request::<()>("heartbeat.status", None).await
    }

    pub async fn heartbeat_start(&self) -> GatewayResult<Value> {
        self.request::<()>("heartbeat.start", None).await
    }

    pub async fn heartbeat_trigger(&self, agent_id: String) -> GatewayResult<Value> {
        if agent_id == "main" {
            self.request::<()>("heartbeat.trigger", None).await
        } else {
            let params = json!({ "agentId": agent_id });
            self.request("heartbeat.trigger", Some(params)).await
        }
    }

    pub async fn pair_list(&self) -> GatewayResult<Value> {
        self.request::<()>("pair.list", None).await
    }

    pub async fn pair_approve(&self, channel: String, sender_id: String) -> GatewayResult<Value> {
        self.request(
            "pair.approve",
            Some(json!({
                "channel": channel,
                "senderId": sender_id,
            })),
        )
        .await
    }

    pub async fn pair_reject(&self, channel: String, sender_id: String) -> GatewayResult<Value> {
        self.request(
            "pair.reject",
            Some(json!({
                "channel": channel,
                "senderId": sender_id,
            })),
        )
        .await
    }

    pub async fn channels_list(&self) -> GatewayResult<Value> {
        self.request::<()>("channels.list", None).await
    }

    pub async fn channel_login(&self, channel: String, account_id: String) -> GatewayResult<Value> {
        self.request(
            "channel.login",
            Some(json!({
                "channel": channel,
                "accountId": account_id,
            })),
        )
        .await
    }

    pub async fn channel_status(
        &self,
        channel: String,
        account_id: String,
    ) -> GatewayResult<Value> {
        self.request(
            "channel.status",
            Some(json!({
                "channel": channel,
                "accountId": account_id,
            })),
        )
        .await
    }

    pub async fn channel_logout(
        &self,
        channel: String,
        account_id: String,
    ) -> GatewayResult<Value> {
        self.request(
            "channel.logout",
            Some(json!({
                "channel": channel,
                "accountId": account_id,
            })),
        )
        .await
    }

    pub async fn channel_stop(&self, channel: String, account_id: String) -> GatewayResult<Value> {
        self.request(
            "channel.stop",
            Some(json!({
                "channel": channel,
                "accountId": account_id,
            })),
        )
        .await
    }

    pub async fn channel_start(&self, channel: String, account_id: String) -> GatewayResult<Value> {
        self.request(
            "channel.start",
            Some(json!({
                "channel": channel,
                "accountId": account_id,
            })),
        )
        .await
    }

    pub async fn tools_list(&self) -> GatewayResult<Value> {
        self.request::<()>("tools.list", None).await
    }

    pub async fn tool_invoke(&self, tool: String, args: Value) -> GatewayResult<Value> {
        self.request("tool.invoke", Some(json!({ "tool": tool, "args": args })))
            .await
    }

    pub async fn config_get(&self, path: Option<String>) -> GatewayResult<Value> {
        match path {
            Some(path) => {
                self.request("config.get", Some(json!({ "path": path })))
                    .await
            }
            None => self.request::<()>("config.get", None).await,
        }
    }

    pub async fn config_set(&self, path: String, value: Value) -> GatewayResult<Value> {
        self.request("config.set", Some(json!({ "path": path, "value": value })))
            .await
    }

    pub async fn skills_status(&self, agent_id: String) -> GatewayResult<Value> {
        if agent_id == "main" {
            self.request::<()>("skills.status", None).await
        } else {
            self.request("skills.status", Some(json!({ "agentId": agent_id })))
                .await
        }
    }

    pub async fn skills_update(
        &self,
        agent_id: String,
        force: bool,
        timeout_ms: Option<u64>,
    ) -> GatewayResult<Value> {
        let mut params = Map::new();

        if agent_id != "main" {
            params.insert("agentId".to_string(), json!(agent_id));
        }

        if force {
            params.insert("force".to_string(), json!(true));
        }

        if let Some(timeout_ms) = timeout_ms {
            params.insert("timeoutMs".to_string(), json!(timeout_ms));
        }

        if params.is_empty() {
            self.request::<()>("skills.update", None).await
        } else {
            self.request("skills.update", Some(params)).await
        }
    }

    pub async fn sessions_list(&self, limit: i64) -> GatewayResult<Value> {
        self.request("sessions.list", Some(json!({ "limit": limit })))
            .await
    }

    pub async fn session_reset(&self, session_key: String) -> GatewayResult<Value> {
        self.request("session.reset", Some(json!({ "sessionKey": session_key })))
            .await
    }

    pub async fn session_get(&self, session_key: String) -> GatewayResult<Value> {
        self.request("session.get", Some(json!({ "sessionKey": session_key })))
            .await
    }

    pub async fn session_stats(&self, session_key: String) -> GatewayResult<Value> {
        self.request("session.stats", Some(json!({ "sessionKey": session_key })))
            .await
    }

    pub async fn session_patch(&self, patch: Value) -> GatewayResult<Value> {
        self.request("session.patch", Some(patch)).await
    }

    pub async fn session_compact(&self, session_key: String, keep: i64) -> GatewayResult<Value> {
        self.request(
            "session.compact",
            Some(json!({ "sessionKey": session_key, "keepMessages": keep })),
        )
        .await
    }

    pub async fn session_history(&self, session_key: String) -> GatewayResult<Value> {
        self.request(
            "session.history",
            Some(json!({ "sessionKey": session_key })),
        )
        .await
    }

    pub async fn session_preview(
        &self,
        session_key: String,
        limit: Option<i64>,
    ) -> GatewayResult<Value> {
        if let Some(limit) = limit {
            self.request(
                "session.preview",
                Some(json!({ "sessionKey": session_key, "limit": limit })),
            )
            .await
        } else {
            self.request(
                "session.preview",
                Some(json!({ "sessionKey": session_key })),
            )
            .await
        }
    }

    pub async fn chat_send(&self, session_key: String, message: String) -> GatewayResult<Value> {
        let run_id = uuid::Uuid::new_v4().to_string();
        self.request(
            "chat.send",
            Some(json!({
                "sessionKey": session_key,
                "message": message,
                "runId": run_id,
            })),
        )
        .await
    }
}
